comp/logs/agent/config/endpoints.go

```diff
@@ -238,6 +267,52 @@ func (e *Endpoint) GetStatus(prefix string, useHTTP bool) string {
 	return fmt.Sprintf("%sSending %s logs in %s to %s on port %d", prefix, compression, protocol, host, port)
 }
 
+// onConfigUpdate handles configuration change notification to update the internal API key of the Endpoint if needed
+func (e *Endpoint) onConfigUpdate(l *LogsConfigKeys) {
+	l.getConfig().OnUpdate(func(key string, oldVal interface{}, newVal interface{}) {
+		if key != e.configSettingPath {
+			return
+		}
+
+		// main Endpoints can directly get their API key from the configuration without having to load complex
+		// types.
+		if !e.isAdditionalEndpoint {
+			if newAPIKey, ok := newVal.(string); !ok {
+				log.Errorf("new API key for '%s' is invalid (not a string) ignoring new value", e.configSettingPath)
+			} else {
+				if oldKey, ok := oldVal.(string); ok && oldKey != e.apiKey.Load() {
+					// This should never happens as it means that an update from the config was
+					// missed
+					log.Warnf("old API key for '%s' doesn't match the one in this endpoints", e.configSettingPath)
+				}
+				log.Infof("rotating API key for '%s': %s -> %s",
+					e.configSettingPath,
+					scrubber.HideKeyExceptLastFiveChars(e.apiKey.Load()),
+					scrubber.HideKeyExceptLastFiveChars(newAPIKey),
+				)
+				e.apiKey.Store(newAPIKey)
+			}
+			return
+		}
+
+		newAdditionalEndpoints, _ := l.getAdditionalEndpoints()
+		if len(newAdditionalEndpoints) <= e.additionalEndpointsIdx {
+			// this should never happens: the number of additional_endpoints should never change at runtime
+			log.Errorf("error: the number of additional_endpoints changed at runtime for '%s', discarding update.", e.configSettingPath)
+			return
+		}
+
+		newAPIKey := newAdditionalEndpoints[e.additionalEndpointsIdx].APIKey
```

Comment:
This is reading the APIKey from the config which has the newly set value, correct? What is the value of `newVal` in cases like this, are we sure it can't be used here?

Reply:
In this case the key would be `logs_config.additional_endpoints` and the value the entire list of additional endpoints.
We could use `newVal` here but that would mean manually casting and traversing it while `getAdditionalEndpoints` already does this.

Reply:
Okay, that makes sense.